[Polyfill] A new component for portability across PHP versions and extensions

nicolas-grekas opened the pull request:
| Q | A |
| --- | --- |
| Bug fix? | no |
| New feature? | yes |
| BC breaks? | no |
| Deprecations? | yes |
| Tests pass? | yes |
| Fixed tickets | - |
| License | MIT |
| Doc PR | - |

I know I come a bit late in the 2.8 releasing process, but I really hope we could merge this new component into 2.8, especially since merging it in 3.1 would remove a great deal of its value since the lowest PHP version there is 5.5, whereas this code has shims for pre-5.5 features.

So here we are:
- polyfills for the `mbstring`
- polyfills for the `Normalizer` class and `grapheme_*` functions
- polyfills for some constants and functions introduced in 5.3, 5.4, 5.5, etc. borrowed from https://github.com/nicolas-grekas/Patchwork/tree/master/core/compat (yet an other code of mine)
- polyfills for `hash_equals`, `hash_pbkdf2` and `ldap_escape` moved out of existing private methods
- utility class `Binary` to be used when `mbstring.func_overload` is enabled

This allows simplifying the full code base by removing many `defined` and `function_exists` checks.

For reference, here is the lists of e.g. functions introduced in 5.4:
http://www.php.net/migration54.functions.php

nicolas-grekas committed:
[Polyfill] A new component for portability across PHP versions and extensions

stof commented on src/Symfony/Component/Polyfill/bootstrap.php:
```diff
@@ -0,0 +1,177 @@
+    define('JSON_BIGINT_AS_STRING', 2);
+    define('JSON_UNESCAPED_SLASHES', 64);
+    define('JSON_PRETTY_PRINT', 128);
+    define('JSON_UNESCAPED_UNICODE', 256);
```
-1 for this. This breaks libraries detecting whether the feature is detected (old versions of Symfony were defining the constant at the beginning of the file, and it was changed precisely for this reason).

constants should be defined only if their behavior is also polyfilled.

stof replied:
the same is true for the JsonSerializable polyfill btw

stof commented:
I don't think creating a component with many polyfills in it is a good idea. Someone wanting to have a ``random_bytes`` should not have to depend on the whole UTF-8 polyfills (especially given that such polyfills cannot be loaded on demand as they are for functions and constants, not for autoloadable classes)

So my vote is a -1 here.

If the ``random_bytes`` polyfill is buggy on Windows, send a PR to the existing library to improve the Windows support, and it will benefit everyone.

nicolas-grekas commented:
This needs some explanations: polyfill **are** autoloaded, they don't pollute anyone’s app.
This is possible thanks to the static classes proxies for polyfills: the bootstrap file is all that is required, with very minimal code to parse/load. Loading the implementation happens only when a polyfill is actually used.

I do think that one single component for all polyfills is a greater idea: polyfills have a finite upper limit (100% feature backporting). By collecting all efforts in one code base, we can concentrate efforts for eveyone, instead of scattering one polyfill here, an other there, with varying implementation quality.

For random_bytes, my implementation is radically different so patching would mean removing most of the code. I'm not that rude.

Pleas reconsider your vote, or at least allow some discussion.

stof commented on src/Symfony/Component/HttpFoundation/Response.php:
```diff
@@ -1147,7 +1147,7 @@ public static function closeOutputBuffers($targetLevel, $flush)
         $status = ob_get_status(true);
         $level = count($status);
-        $flags = defined('PHP_OUTPUT_HANDLER_REMOVABLE') ? PHP_OUTPUT_HANDLER_REMOVABLE | ($flush ? PHP_OUTPUT_HANDLER_FLUSHABLE : PHP_OUTPUT_HANDLER_CLEANABLE) : -1;
+        $flags = PHP_OUTPUT_HANDLER_REMOVABLE | ($flush ? PHP_OUTPUT_HANDLER_FLUSHABLE : PHP_OUTPUT_HANDLER_CLEANABLE);
```
this is actually changing the behavior on PHP 5.3

stof commented on src/Symfony/Component/Polyfill/Php55.php:
```diff
@@ -0,0 +1,99 @@
+        return self::opcache_reset();
+    }
+
+    public static function opcache_reset()
```
polyfilling opcache methods by applying changes in other accelerators looks really weird to me.

stof commented on src/Symfony/Component/Polyfill/Php70.php:
```diff
@@ -0,0 +1,65 @@
+            }
+            $allBytes .= $bytes;
+        }
+        if ($h = @fopen('/dev/urandom', 'rb')) {
```
isn't your implementation affected by https://github.com/paragonie/random_compat/issues/49 ?

nicolas-grekas replied:
true, fixed by excluding windows and checking for char device

nicolas-grekas commented:
I removed constants definitions that did not match any shim implementation. I also added `function_exists` checks to the bootstrap file to enhance compat with existing shims.

stof commented on src/Symfony/Component/Polyfill/bootstrap.php:
```diff
@@ -0,0 +1,162 @@
+    if (!ini_get('session.entropy_file') && !ini_get('session.entropy_length')) {
+        ini_set('session.entropy_length', 32);
+
+        if (false !== @file_get_contents('/dev/urandom', false, null, -1, 1) ) {
```
windows here too

nicolas-grekas replied:
in fact I'm going to remove these ini_setting: it's reducing the entropy for auto conf, not a good

mickaelandrieu commented:
@nicolas-grekas do you plan to make this component available also for Symfony 3 ?

stof commented on src/Symfony/Component/Polyfill/bootstrap.php:
```diff
@@ -0,0 +1,162 @@
+    function mb_http_input($type = '') {return p\Mbstring::mb_http_input($type);}
+}
+
+if (!function_exists('utf8_encode')) {
```
when is it not defined ?

nicolas-grekas replied:
this function is provided by the xml extension, which is not always loaded

stof commented:
@nicolas-grekas discussion is allowed (this is even required to make a decider reconsider their vote btw).

stof commented on src/Symfony/Component/Polyfill/Resources/stubs/Throwable.php:
```diff
@@ -0,0 +1,13 @@
+<?php
+
+interface Throwable
```
Defining this interface while Exception does not implement it does not seem like a good idea to me

nicolas-grekas replied:
fine by me, I'm removing the PHP7 exceptions related shims

stof commented on src/Symfony/Component/Polyfill/Resources/stubs/Error.php:
```diff
@@ -0,0 +1,8 @@
+<?php
+
+class Error extends Exception implements Throwable
```
this stub is misleading. ``Error`` in PHP 7 is *not* catched by a ``catch(\Exception $e)``, while it will catch your stub. this makes the stub unusable

mickaelandrieu commented:
Humm this component is a good idea for Symfony 2.

My only problem is that all components concerned have a new dependency on it, and this component will be - in most cases - useless because PHP 5.5 will become the standard for Symfony 3.

So, if the pull request is only for Symfony 2.8 it's :+1:  else :-1:

stof commented on src/Symfony/Component/Polyfill/Tests/IconvTest.php:
```diff
@@ -0,0 +1,159 @@
+    /**
+     * @covers Symfony\Component\Polyfill\Iconv::iconv_strlen
+     * @covers Symfony\Component\Polyfill\Iconv::strlen1
+     * @covers Symfony\Component\Polyfill\Iconv::strlen2
```
we don't use ``@covers`` in Symfony

nicolas-grekas replied:
yes I know but this imports a code that existed previously, and removing the `@covers` annotation would only reduce the quality of the code (when running coverage tests). Which means I'd like we keep them.

stof commented on src/Symfony/Component/Polyfill/Tests/IntlTest.php:
```diff
@@ -0,0 +1,181 @@
+            $this->assertTrue(true, 'Regular PHP throws a warning');
+        } catch (\PHPUnit_Framework_Error_Notice $e) {
+            $this->assertTrue(true, 'HHVM throws a notice');
+        }
```
shouldn't it actually use a different expectation based on HHVM vs PHP ? There is nothing ensuring that the stub produces a notice on HHVM here

stof commented on src/Symfony/Component/Polyfill/Tests/MbstringTest.php:
```diff
@@ -0,0 +1,275 @@
+ * @author Nicolas Grekas <p@tchwork.com>
+ *
+ * @covers Symfony\Component\Polyfill\Mbstring::<!public>
+ * @requires extension mbstring
```
several tests don't require the extension actually

nicolas-grekas replied:
well, I thought about that and I think it's better for tests to run against reference implementations (because we don't test the shim). So IMO it's better to keep the annotations.

nicolas-grekas commented:
@mickaelandrieu yes I'd like this component to land in 3.0. But of course, all shims for pre-5.5 features should be removed when merging.

mickaelandrieu commented:
@nicolas-grekas ok then :+1: btw great job as always :)

csarrazi commented:
@nicolas-grekas Nice one. However, should this PR limit itself to functions, or also define constants if they are not available?

I'm talking about constants like `LDAP_ESCAPE_DN` or `LDAP_ESCAPE_FILTER`, for example.

nicolas-grekas commented:
@stof comments addressed, tests should be green on travis and appveyor
@csarrazi these constants are required since they are provided as arguments to the ldap_escape() shim.

dunglas commented:
Shouldn't the new polyfill component be only in the `suggest` section of other components but included with the standard edition? It will allow to avoid to force downloading this extra set of files for projects using a recent PHP version.

webmozart commented on src/Symfony/Component/Polyfill/Normalizer.php:
```diff
@@ -0,0 +1,290 @@
+ *
+ * @internal
+ */
+class Normalizer
```
Doesn't this rather belong into the Intl extension? Or should the relevant Intl code be moved here at some point?

nicolas-grekas replied:
Merging Intl's functions.php with bootstrap.php would be great as it will remove one include yes.

webmozart commented:
I like the idea behind this! I agree that one single code base for all polyfills is probably better than several repositories where it's hard to know what is where.

mickaelandrieu commented:
@webmozart agree to your last comment: this is a  dependency I can accept because make theses features available for each component is a mess to maintain and will result to a lot of duplications. 

I'd like to see a pull request opened soon to remove all unused stuff for Symfony 3, I can help you @nicolas-grekas if you don't have time for it now :)

mickaelandrieu commented:
@dunglas what did you suggest? Afaik we can't suggest packages in composer accordingly to the PHP version (could be a great feature indeed)

csarrazi commented on src/Symfony/Component/Polyfill/bootstrap.php:
```diff
@@ -0,0 +1,158 @@
+    if (!function_exists('hash_equals')) {
+        function hash_equals($knownString, $userInput) {return p\Php56::hash_equals($knownString, $userInput);}
+    }
+    if (!function_exists('ldap_escape')) {
```
We should not only check whether the `ldap_escape` function exists or not, but also check whether the extension is enabled or not.

dunglas commented:
@mickaelandrieu if someone needs the Polyfill because he uses an old PHP version it should install the Polyfill Component manually. That way people running on the last PHP version are not forced to download it.

stof commented:
@dunglas the polyfill is not only about older versions, but also about some extensions you don't have.

If we don't require it, we cannot safely use ``mb_*`` functions in components, as it would force users of the component to install either mbstring or the polyfill component, while the code currently works for them all the time. So it would be a BC break.

dunglas commented:
@stof you're right I've not thought about the BC break it introduces.
Maybe can it can be kept in `require` for 2.8 (BC) but removed in 3.0?

stof commented:
@dunglas the experience would still be bad for all these component, as they would be broken by default (and many projects are depending on them, so the end user might not know about the need to choose between installing mbstring or the polyfill component)

csarrazi commented:
@dunglas @stof @nicolas-grekas What about *also* adding a `suggest` for the different PHP extensions, in all bundles which require them?

dunglas commented:
As you want but if the polyfill is always included it would be nice to have an optin notice that upgrading to a newer PHP version or installing an extension would enhance performance and sometimes security (`hash_equals` is less reliable in PHP than in C).

mickaelandrieu commented:
@csarrazi  :+1:

nicolas-grekas commented:
Upgrade pushed with a major enhancement: tests are now run twice automatically. The first time against the internal implementation if available, the second time on the shim implementation. This makes this component a high quality framework for creating shims in PHP (and allowed me to fix some differences already).

I also merged intl's functions.php into bootstrap.php, which means one less global require for composer.
And added the extensions to the composer.json's suggestions.
@webmozart I'd like to keep Normalizer out of Intl for at least one historical reason: I plan to build tchwork/utf8 2.0 on top of this polyfill, but don't want to import 11Mb of data when doing so...

ping @symfony/deciders, all comments addressed, no more work to be done here on my side.

nicolas-grekas commented:
I removed all `@requires extension mbstring|iconv`, tests are still green (see appveyor with `php.ini-min`).

sstok commented:
Is there a special reason for changing the order in random_bytes?

The paragonie/random_compat checks sodium first and then mcrypt while in the Polyfill you changed the this order. And openssl_random_pseudo_bytes() should be considered last as it's actually not fully cryptographically safe! 

/cc @paragonie-scott

nicolas-grekas commented:
@sstok the reason is that it is documented that mcrypt "does the right thing" on Windows an on *nixes.
Since this is a shim, we shouldn't to be better than the internal implementation in PHP7, which is considered good enough.
For openssl, it's better than others when $cryptoSafe is true, which is checked here.

xabbuh commented:
In general, I like this idea. However, I do not yet understand why we not to provide our own implementation of the random functions. Can you explain a bit what would be better with our own implementation compared to the `paragonie/random_compat` package?

nicolas-grekas commented:
@sstok openssl_random_pseudo_bytes is now after `/dev/urandom` and trailing bytes >= length are xored so that their entropy is not lost.
@xabbuh there are several important differences with `paragonie/random_compat`:
- paragonie's adds several (>=5) `require` to the global bootstrapping process, whereas this components adds *none* when the function is not used
- paragonie's loads a shim for `Throwable`&co., which should be out of scope of a shim for random_compat. This adds its own set of issues (see e.g. https://github.com/symfony/symfony/pull/16240#discussion_r42014936)
- paragonie's uses different ways to get entropy in an exclusive way, which means that if the one selected method fails, the full call fails. This issue is happening right now on appveyor and is the cause of our currents failures on 2.8 (looks like mcrypt_create_iv *can* fail)
- paragonie's can use `new COM('CAPICOM.Utilities.1')` on Windows, but does not check that the COM object effectively exists (which is not the case on e.g. appveyor)
- paragonie's uses straight in-place `function random_bytes` declarations, which make it unable to be loaded, thus tested, alongside with the native implementation.

Several of this issues could be fixed with a significant re-factorization but even if that happens, in the end, at least one extra  `require` will remain compared to the implementation in this PR.

paragonie-scott commented:
> For openssl, it's better than others when $cryptoSafe is true, which is checked here.

https://bugs.php.net/bug.php?id=70014

> paragonie's adds several (>=5) `require` to the global bootstrapping process, whereas this components adds none when the function is not used
...
> paragonie's uses different ways to get entropy in an exclusive way, which means that if the one selected method fails, the full call fails. This issue is happening right now on appveyor and is the cause of our current failures on 2.8 (looks like mcrypt_create_iv can fail)

The original design would keep trying CSPRNGs (from the best available to your platform to the worst), but there were two problems here:

1. It would be slower checking every invalid CSPRNG on every invocation of `random_bytes()` and consequently `random_int()`.
2. Constantly falling back to a less-reliable CSPRNG interface is, in one sense, a downgrade attack (though the severity of which is hard to formally express).

Our code determines which is the most secure (based on reasoning provided by ERRATA.md) interface available, and relies on that one interface exclusively in a fail-closed manner. Good security engineering practices encompass how your code handles failure. Silently falling back to an insecure PRNG would be unambiguously a bad move (see: https://github.com/paragonie/random_compat/issues/61 for more). 

Silently falling back to a less-preferable CSPRNG is obviously not as bad, but it's in the same vein, so we don't do that.

> paragonie's can use `new COM('CAPICOM.Utilities.1')` on Windows, but does not check that the COM object effectively exists (which is not the case on e.g. appveyor)

You're right, we should add more checks to the Windows code. I'll look into this today and probably work on getting a new release out early this week.

> paragonie's uses straight in-place function `random_bytes` declarations, which make it unable to be loaded, thus tested, alongside with the native implementation.

Correct. I can easily remedy that for the use-case of unit testing.

paragonie-scott commented:
On another note, I started a similar effort a long time ago to provide a general-purpose polyfill: https://github.com/sarciszewski/php-future

I'm not sure if you want to compare/contrast the two, and I haven't had a chance to audit this PR and compare it to that old effort of mine, but this is the kind of feature that interests me. Let's compare notes and see if we can create a universal polyfill (even if it mostly adds `require` statements to `composer.json`) so even non-Symfony users can benefit from it.

paragonie-scott commented on src/Symfony/Component/Yaml/Parser.php:
```diff
@@ -62,7 +62,7 @@ public function parse($value, $exceptionOnInvalidType = false, $objectSupport =
         $this->lines = explode("\n", $value);
 
-        if (function_exists('mb_internal_encoding') && ((int) ini_get('mbstring.func_overload')) & 2) {
+        if (2 & (int) ini_get('mbstring.func_overload')) {
```
It would be easier for auditing purposes to use `MB_OVERLOAD_STRING` instead of `2` in these instances.

nicolas-grekas replied:
it's not possible because `MB_OVERLOAD_STRING` won't be always defined here...

fabpot replied:
What about doing something like `if (2 /* MB_OVERLOAD_STRING */ & (int) ini_get('mbstring.func_overload')) {`

nicolas-grekas replied:
done

paragonie-scott commented on src/Symfony/Component/Polyfill/Functions/Php70.php:
```diff
@@ -0,0 +1,74 @@
+    public static function random_bytes($length)
+    {
+        if (0 >= $length) {
+            // PHP 7.0 throws an \Error. E_USER_ERROR is the best we can do here.
```
This behavior diverges from what PHP 7 does. At the very least, throw an Exception so it can be wrapped in a `try/catch` block should a developer wish to handle the failure in their own way. My worry here is that raising an `E_ERROR` takes the power out of the user's hands and risks people getting used to the way this polyfill behaves, which would add friction to adopting PHP 7.

paragonie-scott commented on src/Symfony/Component/Polyfill/Functions/Php70.php:
```diff
@@ -0,0 +1,74 @@
+            }
+            $bytes = fread($h, $length);
+            $stat = fstat($h);
+            fclose($h);
```
Continually opening and closing file descriptors is going to add an unnecessary performance penalty.

paragonie-scott commented on src/Symfony/Component/Polyfill/Functions/Php70.php:
```diff
@@ -0,0 +1,74 @@
+            $bytes = fread($h, $length);
+            $stat = fstat($h);
+            fclose($h);
+            if (isset($bytes[$length - 1]) && ($stat['mode'] & 0170000) === 020000) {
```
:+1: for adopting the same sanity checks that @jedisct1 and @narfbg provided random_compat.

paragonie-scott commented on src/Symfony/Component/Polyfill/Functions/Php70.php:
```diff
@@ -0,0 +1,74 @@
+        }
+        $bytes = '';
+        for ($i = 0; $i < $length; ++$i) {
+            $bytes .= chr(mt_rand(0, 255));
```
Do you *really* want to fail open here?

paragonie-scott commented:
So, there are a couple of things I don't like about this:

1. There is no `random_int()`
2. Your `random_bytes()` will silently fallback to `mt_rand()`, which no CSPRNG ever should.

I'd vote :-1: on the code as-is.

paragonie-scott commented:
Also: https://github.com/nicolas-grekas/symfony/blob/f7c98e7d39f9837f25b6d7c605dd23e57ce28531/src/Symfony/Component/Polyfill/Functions/Php70.php#L64-L66

This looks risky.

michaelcullum commented:
Could `intdiv()` also be added?

nicolas-grekas commented:
@paragonie-scott thanks for the helpful review. I updated the code to take your comments into account. Should be ok now. More comments welcomed if you have more.
`random_int` could be provided later so it's not a blocker for now I believe, especially since we don't use it in Symfony.
When this PR gets merged (I hope it will), we should look into https://github.com/sarciszewski/php-future if we can further enhance the component.

paragonie-scott commented:
Well, https://github.com/paragonie/random_compat 1.0.8 is out now, which shouldn't fail on Windows as hard. Thanks for letting me know about these problems.

If there are other issues that arise, please don't hesitate to let me know about them. 

> `random_int` could be provided later so it's not a blocker for now I believe, especially since we don't use it in Symfony.

Are there any compelling reasons to remove random_compat from Symfony, as of 1.0.8?

Take a close, careful look at how `random_int()` is implemented here: https://github.com/paragonie/random_compat/blob/master/lib/random_int.php

ircmaxell commented:
I would **strongly** recommend against including security sensitive polyfills here (random_*, hash_*, etc).

Part of this duplication defeats the point of having canonical implementation. That there is a single, reviewed implementation that can be fixed if a bug is found. The more duplication we have, the more of a chance that a critical issue will be introduced in one of them (and the harder it will be to fix them).

As far as the justifications provided for not using random_compat, I'm not sure there's much weight there (I disagree with most of them outright, but that's just one opinion). Even if there is, provide fixes to that library rather than just reinventing it. 

Oh, and `random_int` is an absolute necessity. Providing one without the other is only going to cause problems as it prevents other polyfills from being installed and working properly.

nicolas-grekas commented:
I removed random_bytes from this PR to keep the discussion focused on the "easy" parts of the proposed polyfill components.

I also added polyfills for `intdiv`, `preg_replace_callback_array` and `*Error` exceptions.

For reference, I submitted https://github.com/paragonie/random_compat/pull/66, https://github.com/paragonie/random_compat/issues/67 and https://github.com/paragonie/random_compat/issues/68 to keep track of the main issues I still have with `paragonie/random_compat`.

nicolas-grekas commented:
README updated

mickaelandrieu commented:
@nicolas-grekas do you think we should update the SymfonyRequirements class from Standard Edition  ?  Some recommandations may be improved: (https://github.com/symfony/symfony-standard/blob/2.8/app/SymfonyRequirements.php#L610).

For instance "Install the mbstring extension to have a better experience" instead of "mb_strlen() should be available".

Could be nice later to provide a more extensions polyfills to reduce the requirements of the framework :cat:

nicolas-grekas commented:
@mickaelandrieu this may be considered after merge, but let's focus on this PR for now.

derrabus commented:
I really like how this Polyfill component makes code easier to read as it removes all the compatibility checks from the logic. It also reduces code duplications by concentrating the shims.

What bloats the component is mainly the charset data for the iconv shim. We more or less only need iconv as a dependency for the mbstring shim (okay, VarDumper calls iconv functions, but that's it, I think). I personally have rarely encountered php installations without the iconv extension. And if shimming iconv is really necessary, adding [patchwork/utf8](https://github.com/tchwork/utf8) to the dependencies probably does the trick.

Also, if I want to mix and match components of different frameworks, a "one polyfill to rule them all" approach might have unexpected side effects, if an application already has own/other shims in place.

What bothers me the most is that we're going to create an omnipresent component that almost every other component depends on. For instance, until now, HttpFoundation and Serializer have been pretty much minimalistic and self-contained components. I could throw them into existing legacy applications and start improving those by replacing legacy code with these components. Now that 900 KB and 500 KB components depends on a 2 MB polyfill package just to shim `json_last_error_msg`, a function with a size of 12 LOC. That's a large overhead. On top of that, since that function was introduced with PHP 5.5, this additional dependency would be for the legacy-on-release 2.8 branch only! I would strongly recommend to leave the dependencies at least of of those two components as they are right now.

teohhanhui commented:
We must all be excited that polyfills are starting to gain prominence in the PHP world, like they have long been in the JS world. But this component really needs to be split up into separate libraries for each type of polyfill, in a style similar to `random_compat`, and from there they can perhaps be bundled together to easily bring it into Symfony Standard Edition.

nicolas-grekas commented:
I was waiting for someone to raise the concern about the iconv shim before proposing to split it in its own sub component :-) I think it's a valid concern that a two precedents: symfony/intl as a partial but weighty shim, and symfony/security-* for subcomponents. It's true also that having iconv not installed should be very exceptional.

derrabus commented:
Setups without ext-iconv should be rather exotic since it's an extension that is enabled by default. To give you another precedent, we're not shimming ext-json either, although you _could_ disable it when compiling php. Maybe we should consider adding ext-iconv as a requirement for 3.0 instead? If, for what reason ever, an application must be deployable in environments without ext-iconv, the developer may still add a compatible shim to their composer.json.

Apart from that, I'm not entirely sure, if a generic web framework like Symfony should maintain a userland implementation of a extensive library like iconv.

And what about my suggestion of not shimming `json_last_error_msg` to keep the dependencies of low-level components clean, @nicolas-grekas?

mickaelandrieu commented:
@teohhanhui agree but this is not the Symfony way to manage packages.  This component will be part of symfony/symfony  metapackage and then all "standalone" components are splitter.  From a maintainer point of view,  maintain all theses poly fill is hard and without no real effective gain in terms of performance.

But if it's possible and easy to maintain,  split this Polyfill into subcomponents php55, php56, php7 and iconv would be great!

nicolas-grekas commented:
@mickaelandrieu This component is about having one main repository for high quality shims. Splitting it into separate sub components is not desired. We already tried with symfony/intl in the past and it proved being a mistake, by significantly complicating composer.lock files management. Saving a few kb in package size is not a target esp. since there is no measurable perf impact in adding more files on the fs.

@derrabus for the same reasons, I see no value into trying to not depend on the shim for http-foundation. On the contrary, delegating legacy concerns to the Polyfill allows tranfering the technical debt of http-foundation. For 3.0, since the lowest php req has been raised, the dep might be removed thought.

For iconv, what about the following plan: put the shim into v2.8 and remove it from 3.0? This makes it possible to select the trimmed down version (no iconv) just by playing with the composer requirements. It would also pave the way to adding iconv as a requirement to Sf 3.0, which is, as already said, not a stronger req than the existing one for ext-json.

nicolas-grekas commented:
I removed the iconv polyfill in the hope that https://github.com/symfony/symfony-standard/issues/872 will be accepted

derrabus commented on src/Symfony/Bundle/FrameworkBundle/Command/TranslationDebugCommand.php:
```diff
@@ -240,7 +240,7 @@ private function sanitizeString($string, $length = 40)
         $string = trim(preg_replace('/\s+/', ' ', $string));
 
-        if (function_exists('mb_strlen') && false !== $encoding = mb_detect_encoding($string)) {
+        if (false !== $encoding = mb_detect_encoding($string)) {
```
This change makes Polyfill a dependency of FrameworkBundle, doesn't it?

nicolas-grekas replied:
indeed, fixed

paragonie-scott commented on src/Symfony/Component/Security/Csrf/composer.json:
```diff
@@ -17,8 +17,8 @@
     "require": {
         "php": ">=5.3.9",
-        "symfony/security-core": "~2.4|~3.0.0",
-        "paragonie/random_compat": "~1.0"
```
Is removing random_compat from the CSRF module intentional? In most situations this should already be available through other components.

nicolas-grekas replied:
no it's not, fixed, thanks

paragonie-scott commented:
Reviewed your `hash_pbkdf2()` and `hash_equals()` polyfills, didn't find any bugs. :+1:

fabpot commented on src/Symfony/Component/Polyfill/bootstrap.php:
```diff
@@ -0,0 +1,141 @@
+
+if (PHP_VERSION_ID < 50400) {
+    if (!function_exists('trait_exists')) {
+        function trait_exists($class, $autoload = true) {return $autoload && class_exists($class, $autoload) && false;}
```
Can you add a space before `return` and `}`?

fabpot commented:
:+1: ping @symfony/deciders

rparsi commented on src/Symfony/Component/Polyfill/Normalizer.php:
```diff
@@ -0,0 +1,292 @@
+                $result .= $lastUchr;
+                $lastUchr = $s[$i];
+                ++$i;
+            } else {
```
@nicolas-grekas Hi I'm just a random php developer that uses Symfony so feel free to disregard my comment.
You can remove this 'else' by adding a 'continue' at the end of the sibling 'if' above; visually makes it less nested.
Same for the 'if' on line 130.

rparsi commented on src/Symfony/Component/Polyfill/Normalizer.php:
```diff
@@ -0,0 +1,292 @@
+                $j = 1 + strspn($s, $ASCII, $i + 1);
+                $result .= substr($s, $i, $j);
+                $i += $j;
+            } else {
```
@nicolas-grekas This just me, I think this would be more readable if the 'if' statements had a 'continue' at the end thereby negating the need for the 'else'.

nicolas-grekas replied:
Good idea, done

rparsi commented:
@nicolas-grekas My apologies in advance for my newb comments/questions.

Questions:
1. Does Symfony really need this?  Are there a lot of dependencies (php version, specific extensions) on character encoding functions and random number functions?
2. Shouldn't the symfony installer account for this when creating new project (by modifying initial composer.json/composer.lock)?

Comments:
I think it's better to have separate polyfills for the char encoding and number generation.  Keeping them as one component would most likely cause scope problems, code gets bloated and out of control.

I have concerns as to how this would affect the Symfony component landscape, in that components are supposed to be self contained and independent.  Would core Symfony components now become dependent on the polyfill components?

rparsi commented on src/Symfony/Component/Polyfill/Functions/Mbstring.php:
```diff
@@ -0,0 +1,604 @@
+                $uchr = substr($s, $i, $ulen);
+                $i += $ulen;
+
+                if (isset($map[$uchr])) {
```
@nicolas-grekas You're going to get really annoyed with me by now.
In general I prefer to use 'continue' in 'if' blocks in loops (while/foreach/whatever) so that nesting is reduced, making code easier to read.

nicolas-grekas replied:
That's a way of contributing :-)

rparsi replied:
@nicolas-grekas If you're going to embark on that adventure, I see a lot of loops where you can reduce the nesting.  So if you find yourself cursing while changing the code feel free to document that via github comments.

nicolas-grekas commented:
How adding a dependency could be a concern? There is no measurable downside. But I see the benefits of having one rock solid polyfill. See my comments above.

dunglas commented:
I'm still annoyed by this new component being a dependency of almost all other components even with Symfony 3.0 (as I already explained, it doesn't make sense to download some extra MB for nothing in the case where extensions are already installed and the PHP version is recent) but the component itself makes a lot of sense so :+1: to merge it in 2.8.

derrabus commented:
The dependency we're about to add affects the global namespace. No other Symfony Component did this to such an extent. I fear side effects in projects that only use some few Symfony Components and that might already have other/own polyfills in place. See my concerns regarding the self-contained components HttpFoundation and Serializer above.

teohhanhui commented:
If `Security` was split from one monolithic component into smaller single-purpose components, why can't the same be done here?

nicolas-grekas commented:
@derrabus the intl component already does inject a few functions in the global namespace. In this PR, the only file that injects things into the global namespace is bootstrap.php. We should carefully review it so that we ensure that your concern does not apply. I just did so: all declarations are enclosed in `if` blocks that IMO are granular enough to prevent any collision with any existing shim (that also must be enable conditionally). What can happen is that an implementation from this PR can take over an other existing shim. But since it's a shim, either they behave the same and it's not an issue at all, or they behave differently and we will able to decide the right behavior by comparing to the native implem behavior. And fixing the shim here if needed. This is theoretical pov: all shims provided here are were already written prio being proposed for this component. Which means they are quite proven. In practice, we're talking about a few simple functions that can't diverge from the native behavior, thus are "correct". And the most complex ones (e.g. hash_equals) have already been reviewed carefully in the past.

@teohhanhui it's not possible to conditionally load packages at the composer level, which means it's not possible to split into several package without requiring everyone to take care of this subject by understanding why then adding explicit requirements to the composer.json. Which is btw not possible for OSS when portability is a basic requirement.

@dunglas just to tell real facts "some extra MB" is wrong: 500kb is the right number (80kb compressed)

nicolas-grekas commented:
By leveraging the fact that iconv is now a requirement of sf 2.8, I removed the polyfill dependency for the Debug and Templating components and for the Doctrine bridge also. The others are legitimate.

stof commented:
@nicolas-grekas intl was not split separately. And what caused us issues in the past with symfony/icu was not that it was in a separate package. It was that different versions of the package needed to be installed depending on whether ext-intl was installed or no (meaning it was *not* written as a polyfill).

To make things easier for packages needing a polyfill for only a few functions and not wanting to depend on a huge polyfill, I agree with people saying previously that they should probably be separate packages, allowing composer packages to depend only on the polyfill they really need.

Instead of creating a single Symfony component aimed at containing all high-quality polyfills, what do you think about creating a github organization (and a composer vendor namespace) dedicated at high-quality polyfills instead ? This way, it would be possible to have many separate packages. And Symfony could depend on the polyfills it needs (allowing HttpFoundation and Serializer to depend only on a ``json_last_error_msg`` polyfill, and dropping the dependency in 3.0 too)

teohhanhui commented:
@nicolas-grekas
> it's not possible to conditionally load packages at the composer level, which means it's not possible to split into several packages without requiring everyone to take care of this subject by understanding why then adding explicit requirements to the composer.json. Which is btw not possible for OSS where some portability is a basic requirement.

What I mean is that each Symfony component should only depend on the polyfill packages that it needs. It's not conditional requirements.

nicolas-grekas commented:
> What I mean is that each Symfony component should only depend on the
polyfill packages that it needs. It's not conditional requirements.

OK. Then what benefits do you expect?

nicolas-grekas commented:
@stof I'm sorry but I really don't see what's "huge" in this polyfill. Splitting in into several packages would only a more `require` do the bootstrapping process without any concert benefit. Or put an other way: what would be the benefits of splitting?

rparsi commented on src/Symfony/Component/Polyfill/Functions/Mbstring.php:
```diff
@@ -0,0 +1,604 @@
+                $map = $lower;
+            }
+
+            static $ulenMask = array("\xC0" => 2, "\xD0" => 2, "\xE0" => 3, "\xF0" => 4);
```
@nicolas-grekas Why isn't this a class property?

rparsi commented:
@nicolas-grekas @stof @mickaelandrieu @csarrazi @dunglas @webmozart @sstok @xabbuh @paragonie-scott @ircmaxell @derrabus @teohhanhui @fabpot 
DISCLAIMER: I'm just a random php dev so feel free to ignore this, and my apologies in advance for anything newb-ish.

I've read all the comments and looked at the code changes, and here's my attempt to summarize the concerns so you the maintainers/collaborators of Symfony can decide what to do.

1. The Polyfill component doesn't behave like a component in that other components become dependent on it.  I'm using the current definition of components as defined here http://symfony.com/components
which is "Symfony Components are a set of decoupled and reusable PHP libraries"

2. As others have pointed out, this Polyfill "super" component (as in super class, sorry can't think of a better description atm) introduces compatibility problems for mixing/matching components, as developers may need to for their given use case.

Personally I like stof's idea:
"Instead of creating a single Symfony component aimed at containing all high-quality polyfills, what do you think about creating a github organization (and a composer vendor namespace) dedicated at high-quality polyfills instead ? This way, it would be possible to have many separate packages. And Symfony could depend on the polyfills it needs (allowing HttpFoundation and Serializer to depend only on a json_last_error_msg polyfill, and dropping the dependency in 3.0 too)"

You get the benefits of the polyfills with less headaches.

Just my opinion:
The Polyfill component as is in this PR solves one problem (cleaner code) but breaks the meaning of what a Symonfy Component is supposed to be, and introduces potential compatibility headaches for the rest of us common php developer folks.  As someone who really likes the Symfony framework and it's decoupled nature this is alarming; Polyfill feels like a hack almost.

If you guys decide to go down this road then the official documentation should be updated, so newbs like me aren't mislead.

derrabus commented:
@rparsi :clap:

nicolas-grekas commented:
@rparsi thanks for your input but you misread the previous comments and the definition:
- some raised compat concern, but none is proven, at least nobody refuted my arguments saying there is no compat issue.
- the definition a component does not prevent components being dependent on each others, which is of course already the case for many components. There is no difference.
...

rparsi commented:
@nicolas-grekas @derrabus 
That's why I put the disclaimer first :)

I think I can summarize it better:
The added dependency between Polyfill and other components itself is not the root concern, it's that the scope/depth of the dependency is considerable or extreme compared to the current Symfony framework/ecosystem.

So I think the real vote should be on whether or not to do something as @stof suggests.

By the way derrabus, I can't tell if your clap is sarcastic or earnest. :)

mickaelandrieu commented:
> the definition of a component does not prevent components being dependent on each others, which is of course already the case for many components. There is no difference.

this argument sounds bad to me: this is not because we don't respect the promises we have done before that we have to continue this way.
Since 2.3, more and more components are coupled to anothers - and of course -  this is bad for reuse. Symfony project is too much "Symfony Standard Edition" centric. I stopped fighting against it, doesn't means - like a lot of developers - that I think this is the good way to go.

rparsi commented:
@mickaelandrieu Thanks, you captured my sentiments exactly.

lsmith77 commented on src/Symfony/Component/Polyfill/README.md:
```diff
@@ -0,0 +1,51 @@
+compatibility PHP implementations for some extensions and functions. You should
+use it when portability across PHP versions and extensions is desired.
+
+Polyfills are provided are for:
```
``Polyfills are provided for:``

derrabus commented:
@rparsi earnest

ircmaxell commented:
One question: What does this have to do with Symfony at all? Why isn't this a generic PHP package (why does it have to be a Symfony component, or shipped by Symfony)?

fabpot commented:
@ircmaxell I'm not sure I understand your concern. Symfony2 has been a set of components from day one, so why not having one on polyfills? We already have similar components like the Intl one. Moreover, Symfony is not just this monolith repository as each component has its own separate repository (which is automatically synced in real-time with the main Symfony one) and Composer package. So, in that sense, that's no different from any other "generic PHP package" (not sure I understand what a generic PHP package is to be honest though). And of course, being under the Symfony umbrella brings a lot of "benefits" like LTS releases, the BC promise, documentation, 2 releases a year, and much more. Last, but not the least, it just happens that we need that "abstraction" in Symfony, that much of the code was already scattered in the different Symfony components (sometimes duplicated) and Nicolas, who is working on this topic is a member of the Symfony core team. Seems very natural to me that we are doing that component in the Symfony project itself and not elsewhere. It looks like I'm missing something here. Or does it has a negative connotation when something is part of Symfony? Is it a bad thing?

paragonie-scott commented:
I think @ircmaxell was trying to suggest making one canonical framework-agnostic polyfill for new PHP features rather than having it be something that Symfony provides and maintains for Symfony exclusively, because if every framework does this, we end up with needless code duplication and bugs fixed in one place might not percolate to the rest of the community.

I don't think he meant to imply any negative connotation just because it's a part of Symfony.

fabpot commented:
@paragonie-scott I did understand that. But why the Symfony one couldn't be the canonical one? It would be an standalone Github repository with its own Packagist entry (no other Symfony deps). So, the only reason I can think of is the "Symfony" name. I thought that the PHP community stopped framework wars years ago and that nowadays people were more open to use components/libraries coming from many different frameworks. Or am I wrong?

paragonie-scott commented:
> But why the Symfony one couldn't be the canonical one?

I don't see why it couldn't.

I started an initiative to make [a polyfill](https://github.com/sarciszewski/php-future) a while ago, but it's nowhere near as comprehensive. If `Symfony\Polyfill` is accepted by the community, I have no problem retiring it in favor of this one.

> It would be an standalone Github repository with its own Packagist entry (no other Symfony deps).

That's a good point in favor of this one.

> I thought that the PHP community stopped framework wars years ago and that nowadays people were more open to use components/libraries coming from many different frameworks. Or am I wrong?

Well, this might be off-topic, but I still see pointless arguments about which framework is better almost every day, on Reddit, LinkedIn, Facebook, various forums, and IRC. Very rarely among professionals. It's hard to say if they're really over or not.

Tobion commented:
In general I think the Polyfill component is a good thing. It bundles all those fallback implementations in one place instead of having it spread across different components and each having to deal with workarounds.

I can think of one major argument why should should live outside of symfony/symfony. We increase minimum PHP requirements in symfony with major versions. But for the Polyfill it would not make sense to have the same release timeline and minimum PHP requirement as all the other symfony components.  The Polyfill will probably have to provide the Polyfill for older PHP versions longer than symfony components because that is one of the major points of Polyfill. So it makes no sense to tie Polyfill and Components together and increase minimum PHP requirements at the same time. Otherwise it would mean we have to create a version 3 of Polyfill right now that removes all Polyfills for PHP < 5.5 which makes no sense already...

Tobion commented:
The release timeline was also one argument why we extracted several bundles/components from symfony/symfony in the past. And for Polyfill this is even more relevant IMO.

nicolas-grekas commented:
The 2.8 polyfill is for 5.3.9 minimum, and it will be supported in a LTS release cycle. 5.3 and 5.4 won't get any new feature. Which means the polyfill is feature full. Closed. Over. It won't receive anything more but bug fixes if any. So II there is no issue at all. Target achieved, package published, job done, we can all move on at let the legacy there.
So I really don't see your point...

ircmaxell commented:
> And of course, being under the Symfony umbrella brings a lot of "benefits" like LTS releases, the BC promise, documentation, 2 releases a year, and much more. 

None of which benefit a polyfill. Polyfill versioning needs to be tied to the release cycle of the thing being polyfilled (as 7.1 is coming out, the polyfill would need to be updated for new patches), as well as for new extensions (which come at sporatic times). 

As far as documentation, if it differs from php.net, it's not a polyfill (or that difference is a bug).

> Last, but not the least, it just happens that we need that "abstraction" in Symfony

The difference though that I see is that every other component in Symfony is a relatively opinionated component. It's not aiming to be a canonical implementation, but instead an option. This is different and doesn't fit that paradigm. What we're talking about here is an unopinionated implementation of something that's directly tied to PHP itself. 

It's not an abstraction at all. It's a polyfill. If you wanted to maintain an abstraction that included polyfill *behavior*, then I'd agree 100%. But that's not what's being discussed here. We're discussing a direct polyfill.

Ultimately do what you want, I'm not going to point fingers or anything. I'm just pointing out that this is unlike any of the other Symfony components, and by definition has different requirements (as I point out above in the versioning issue). It just feels "odd" to me. Call me crazy (I know, many do), but I'm just trying to point out that this is unlike anything else that you call a component (and calling it a component actually may work against you).

rparsi commented:
@nicolas-grekas @fabpot From your comments you guys really want to merge this.  I have a request that if implemented should help others detect/debug conflicts with other components (custom or otherwise).

Nicolas how much work would it be to add a console command that outputs the following information:
- your environments php version
- the list of functions that the polyfill thinks it should implement for your application (ie if you triggered a controller action), and for each one if it already exists before the Polyfill would implement it

@ircmaxell Thanks you explain it a lot better then I ever could.  This Polyfill works like a component but as it's tied to PHP itself feels out of place as a Symfony component.  Not sure if I'm making sense here so my apologies.

Tobion commented:
:-1: to include it in `symfony/symfony`
+0 to create a separate repo like `symfony/polyfill` that bundles all polyfills
+1 to create a vendor/organisation just for polyfills where each polyfilled extension has it's own repo as stof suggested

rparsi commented:
I know my vote doesn't count, so this just for myself :)
:+1: to @Tobion @stof

nicolas-grekas commented:
Sleeping helping, I now understand what @Tobion says about minimum PHP versions and why that doesn't fit for a polyfill component.

fabpot commented:
Thank for the constructive discussion everyone, I really appreciate it. We heard your arguments, and we've decided to create a new repo for polyfills, symfony/polyfill which is going to have a different version numbering than Symfony itself and a different release cycle.

Go over at https://github.com/symfony/polyfill for the new implementation which should be ready soon.

edlington commented:
I don't see how polyfills result in loosely coupled code. What I'm seeing is that code rewritten to remove function_exists checks for backwards compatibility now will fail unless polyfill components are used.  Above it says polyfills are autoloaded. They don't seem to be autoloading in my install. One definition of clean code is to minimize the number of conditions and loops in a function. That's great. Creating new components moves complexity from the file level to the directory level. Cleaner code, more confusing directory structure, more difficult to track down errors. For Symfony wizards, no problem. Solutions are always political and in this case I think the polyfills make the framework less friendly for those of us who are learners rather than experts.

derrabus commented:
@edlington If the polyfills don't work for you as expected, please open an issue explaining your problem. This gives others a chance to help you.

stof commented:
@edlington are you using the Composer autoloader, or are you using your own autoloading system ?

edlington commented:
@stof thank you for that question. I think I see the problem is that I need to add an autoload field to my composer file.
